types/event_bus_test.go

```diff
@@ -0,0 +1,122 @@
+package types
+
+import (
+	"context"
+	"fmt"
+	"math/rand"
+	"testing"
+	"time"
+
+	tmpubsub "github.com/tendermint/tmlibs/pubsub"
+)
+
+func BenchmarkEventBus(b *testing.B) {
+	benchmarks := []struct {
+		name        string
+		numClients  int
+		randQueries bool
+		randEvents  bool
+	}{
+		{"10Clients1Query1Event", 10, false, false},
+		{"100Clients", 100, false, false},
+		{"1000Clients", 1000, false, false},
+
+		{"10ClientsRandQueries1Event", 10, true, false},
+		{"100Clients", 100, true, false},
+		{"1000Clients", 1000, true, false},
+
+		{"10ClientsRandQueriesRandEvents", 10, true, true},
+		{"100Clients", 100, true, true},
+		{"1000Clients", 1000, true, true},
+
+		{"10Clients1QueryRandEvents", 10, false, true},
+		{"100Clients", 100, false, true},
+		{"1000Clients", 1000, false, true},
+	}
+
+	for _, bm := range benchmarks {
+		b.Run(bm.name, func(b *testing.B) {
+			benchmarkEventBus(bm.numClients, bm.randQueries, bm.randEvents, b)
+		})
+	}
+}
+
+func benchmarkEventBus(numClients int, randQueries bool, randEvents bool, b *testing.B) {
+	// for random* functions
+	rand.Seed(time.Now().Unix())
+
+	eventBus := NewEventBusWithBufferCapacity(0) // set buffer capacity to 0 so we are not testing cache
+	eventBus.Start()
+	defer eventBus.Stop()
+
+	ctx := context.Background()
+	q := EventQueryNewBlock
+
+	for i := 0; i < numClients; i++ {
+		ch := make(chan interface{})
```

Comment:
Am curious, in the design of our events, when do we this channel? I ask because perhaps instead eventBus.Subscribe should return the channel as a read-only one? That was it can be closed to indicate when no more events are coming through or on error. We can discuss that later though, that shouldn't block your progress on this PR since this work is moving over the old events package.

Reply:
> That was it can be closed to indicate when no more events are coming through or on error.

this is the case right now.

> I ask because perhaps instead eventBus.Subscribe should return the channel as a read-only one? 

I remember going back and forth between the current design and this. The current design allows a subscriber to choose the type of the channel: buffered (we need this in tests) or unbuffered.